### PR TITLE
MONIT-17100: Implemented real-time spans generation and sending.

### PR DESCRIPTION
**Pull request by dbaghdasarya**

### CODE REVIEW

File: src/main/java/com/wavefront/SpanGenerator.java [hunk 1]
```diff
@@ -337,4 +337,53 @@ private int getIndexOfNextItem(@Nonnull List<Integer> percentages) {
...
+      current = System.currentTimeMillis();
+      mustBeGeneratedSpans = (int) (rate * (current - start) / 1000);
+
+      while (generatedSpans < mustBeGeneratedSpans) {
+        // get next trace type to be generated
+        TraceTypePattern traceTypePattern = getNextTraceType(generatorConfig.getTraceTypePatterns());
+        List<List<Span>> trace = generateTrace(traceTypePattern);
+        spanQueue.addTrace(trace);
+        for (List<Span> spanList :
+            trace) {
```
Comment (sirak92) [inline src/main/java/com/wavefront/SpanGenerator.java]: can be in one line. Also if we need the spans count the generateTrace() function can be slightly changed to return the generated spans count as well.

File: src/main/java/com/wavefront/SpanGenerator.java [hunk 2]
```diff
@@ -337,4 +337,53 @@ private int getIndexOfNextItem(@Nonnull List<Integer> percentages) {
...
+    int spansCount;
+    int rate = generatorConfig.getSpansRate();
+    if (traceCount > 0) {
+      LOGGER.info("Should be generated " + traceCount + " traces.");
+      whileCheck = queue -> queue.getEnteredTraceCount() < traceCount;
+    } else {
+      spansCount = rate * (int) generatorConfig.getDuration().toSeconds();
+      LOGGER.info("Should be generated " + spansCount + " spans.");
+      whileCheck = queue -> queue.getEnteredSpanCount() < spansCount;
+    }
```
Comment (sirak92) [inline src/main/java/com/wavefront/SpanGenerator.java]: this part is duplication from generate() function. So here is a question: do we need a generate() function at all? Even it's needed we can encapsulate the generic part in a new function.

File: src/main/java/com/wavefront/SpanQueue.java [hunk 3]
```diff
@@ -2,44 +2,65 @@
...
+        spanQueue.addAll(spans);
+        spanCount.addAndGet(spans.size());
+      });
+    }
+    traceCount.addAndGet(1);
   }
 
-  public int getTraceCount() {
-    return traceCount;
+  public int getEnteredTraceCount() {
```
Comment (sirak92) [inline src/main/java/com/wavefront/SpanQueue.java:34]: maybe rename to getGeneratedTraceCount()? or more intuitive name?

File: src/main/java/com/wavefront/SpanQueue.java [hunk 4]
```diff
@@ -2,44 +2,65 @@
...
+    return traceCount.get();
+  }
+
+  public int getEnteredSpanCount() {
```
Comment (sirak92) [inline src/main/java/com/wavefront/SpanQueue.java:38]: same here

File: src/main/java/com/wavefront/WavefrontTraceLoader.java [hunk 5]
```diff
@@ -17,51 +16,53 @@
...
     new WavefrontTraceLoader().start(args);
   }
 
   @Override
   void setupSenders() throws IOException {
-    ApplicationConfig config = loadApplicationConfig();
-    if (config == null) {
+    if (applicationConfig == null) {
       throw new IOException("Application config should contain proxy or direction ingestion info.");
     }
```
Comment (sirak92) [inline src/main/java/com/wavefront/WavefrontTraceLoader.java:31]: this check is not needed anymore

File: src/main/java/com/wavefront/WavefrontTraceLoader.java [hunk 6]
```diff
@@ -74,4 +75,26 @@ void dumpStatistics() throws Exception {
...
+    sender.join();
+  }
+
+  void generateSpans() {
+    spanGenerator.generate();
+  }
+
+  void saveSpansToFile() throws Exception {
+    spanSender.saveToFile();
+  }
```
Comment (sirak92) [inline src/main/java/com/wavefront/WavefrontTraceLoader.java]: I will suggest getting rid of these one-line functions.

File: src/main/java/com/wavefront/SpanGenerator.java (hunk 2 above)
Comment (dbaghdasarya) [reply src/main/java/com/wavefront/SpanGenerator.java]: Moved the common part to a new method. Also renamed generate() method to generateForFile().

File: src/main/java/com/wavefront/SpanGenerator.java (hunk 1 above)
Comment (dbaghdasarya) [reply src/main/java/com/wavefront/SpanGenerator.java]: Replaced with one line. In my next task, I will create "Trace" class which will provide additional functionality.

File: src/main/java/com/wavefront/SpanQueue.java (hunk 3 above)
Comment (dbaghdasarya) [reply src/main/java/com/wavefront/SpanQueue.java:34]: SpanQueue has no idea where from a trace comes, it only knows that a trace was entered to the queue.

File: src/main/java/com/wavefront/WavefrontTraceLoader.java (hunk 5 above)
Comment (dbaghdasarya) [reply src/main/java/com/wavefront/WavefrontTraceLoader.java:31]: It is possible that applicationConfig will not be instantiated before this point.

File: src/main/java/com/wavefront/WavefrontTraceLoader.java (hunk 6 above)
Comment (dbaghdasarya) [reply src/main/java/com/wavefront/WavefrontTraceLoader.java]: Agree!

**Pull request merged**